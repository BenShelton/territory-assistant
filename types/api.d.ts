import { IInfoText } from './territory'
import { ISettings, ILoginRequest, ILoginResponse } from 'types'
import { UpdatedSettings } from './settings'

export namespace API {
  export namespace Auth {
    export namespace Login {
      export type Request = ILoginRequest
      export type Response = ILoginResponse
    }

    export namespace Logout {
      export type Request = void
      export type Response = true
    }
  }

  export namespace Settings {
    export namespace Load {
      export type Request = void
      export type Response = ISettings
    }
    export namespace Update {
      export type Request = UpdatedSettings
      export type Response = ISettings
    }
  }

  export namespace Info {
    export namespace List {
      export type Response = IInfoText[]
    }
    export namespace Add {
      export type Request = IInfoText
      export type Response = IInfoText
    }

    export namespace Update {
      export type Request = IInfoText
      export type Response = IInfoText
    }

    export namespace Delete {
      export type Request = string
      export type Response = true
    }
  }
}
