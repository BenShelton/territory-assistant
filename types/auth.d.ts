export type JWT = string

export interface IToken {
  _id: string
}

export interface ILoginRequest {
  password: string
}

export interface ILoginResponse {
  token: JWT
}
