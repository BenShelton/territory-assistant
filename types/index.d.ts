export * from './map'
export * from './territory'
