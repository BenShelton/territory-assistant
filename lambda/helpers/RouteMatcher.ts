import { APIGatewayEvent } from 'aws-lambda'

type APIMethod = 'GET' | 'POST' | 'PUT' | 'DELETE'

export class RouteMatcher {
  private _paramRegex = /:(.*?)(?=(\/|$))/
  private _event: APIGatewayEvent
  private _pathParams: Record<string, string> = {}

  constructor (event: APIGatewayEvent) {
    this._event = event
  }

  get pathParams (): Record<string, string> {
    return this._pathParams
  }

  private _getMatches (path: string, regex: RegExp): string[] {
    const matches = path.match(regex)
    if (!matches) return []
    return matches.slice(1)
  }

  testRoute (method: APIMethod, path: string): boolean {
    if (this._event.httpMethod !== method) return false
    const stringParamRegex = new RegExp(path.replace(this._paramRegex, '(\\w+)') + '/?$')
    if (!stringParamRegex.test(this._event.path)) return false
    const paramNames = this._getMatches(path, this._paramRegex)
    const paramValues = this._getMatches(this._event.path, stringParamRegex)
    this._pathParams = paramNames.reduce((acc: Record<string, string>, p, i) => {
      if (p) acc[p] = paramValues[i]
      return acc
    }, {})
    return true
  }
}
