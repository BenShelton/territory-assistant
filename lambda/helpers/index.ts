import { APIGatewayEvent } from 'aws-lambda'

type APIMethod = 'GET' | 'POST'
type RouteMatcher = (method: APIMethod, path: string) => boolean
interface IResult {
  statusCode: number
  body: string
}

export function generateRouteMatcher (event: APIGatewayEvent): RouteMatcher {
  return function routeMatcher (method: APIMethod, path: string) {
    if (event.httpMethod !== method) return false
    return new RegExp(path + '/?$').test(event.path)
  }
}

export function success<T> (data: T): IResult {
  return { statusCode: 200, body: JSON.stringify(data) }
}

export function badRequest (message: string): IResult {
  return { statusCode: 400, body: message }
}

export function notFound (): IResult {
  return { statusCode: 404, body: 'Unknown Route' }
}
