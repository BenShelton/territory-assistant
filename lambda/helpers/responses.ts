interface IResult {
  statusCode: number
  body: string
}

export function success<T> (data: T): IResult {
  return { statusCode: 200, body: JSON.stringify(data) }
}

export function badRequest (message: string): IResult {
  return { statusCode: 400, body: message }
}

export function unauthorized (): IResult {
  return { statusCode: 401, body: 'Unauthorized' }
}

export function notFound (): IResult {
  return { statusCode: 404, body: 'Unknown Route' }
}
