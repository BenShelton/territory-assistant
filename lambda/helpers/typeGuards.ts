import { IPoint } from 'types'

export function isObject (obj: unknown): obj is Record<string, unknown> {
  return typeof obj === 'object' && !Array.isArray(obj) && !!obj
}

export function isPointList (obj: unknown): obj is IPoint[] {
  if (!Array.isArray(obj)) return false
  const basePoint: IPoint = { lat: 0, lng: 0 }
  return Object.entries(basePoint)
    .every(([k, v]) => {
      return obj.every(p => {
        if (!isObject(p)) return false
        return typeof p[k] === typeof v
      })
    })
}
