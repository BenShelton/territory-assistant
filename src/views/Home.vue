<template>
  <v-container
    flex-column
    flex-nowrap
    align-start
    fill-height
  >
    <h1 class="headline mb-3">
      Welcome to Territory Assistant
    </h1>
    <TerritoryEditor
      :active-layers="['image', 'territory', 'maps', 'info']"
    />
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue'

import TerritoryEditor from '@/components/TerritoryEditor.vue'

export default Vue.extend({
  name: 'Home',

  components: {
    TerritoryEditor
  }
})
</script>
