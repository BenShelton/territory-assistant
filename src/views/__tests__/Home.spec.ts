import { createWrapper } from 'UnitTest/helpers'

import Home from '../Home.vue'

import { API } from 'types'

describe('Views/Home', () => {
  // Setup
  beforeAll(() => {
    fetchMock.mockResponse(async (req) => {
      const url = req.url.replace('/.netlify/functions/', '')
      switch (url) {
        case 'info': {
          const res: API.Info.List.Response = []
          return JSON.stringify(res)
        }
      }
      return JSON.stringify({})
    })
  })

  // Tests
  it('should have the name "Home"', () => {
    const wrapper = createWrapper(Home, { store: true })
    expect(wrapper.name()).toBe('Home')
  })

  it('should match the snapshot', () => {
    const wrapper = createWrapper(Home, { store: true })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
