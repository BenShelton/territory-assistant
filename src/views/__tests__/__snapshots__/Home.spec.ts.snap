// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Views/Home should match the snapshot 1`] = `
<div class="container flex-column flex-nowrap align-start fill-height">
  <h1 class="headline mb-3">
    Welcome to Territory Assistant
  </h1>
  <div class="editor v-card v-sheet theme--light">
    <!---->
    <div role="dialog" class="v-dialog__container">
      <div role="document" class="v-dialog__content" style="z-index: 0;">
        <div class="v-dialog" style="max-width: 300px; display: none;"></div>
      </div>
    </div>
    <!---->
    <div id="map"></div>
  </div>
</div>
`;
